pkg/nodejs/yarn/parse.go

```diff
@@ -69,60 +136,110 @@ func NewParser() types.Parser {
 func (p *Parser) Parse(r dio.ReadSeekerAt) (libs []types.Library, deps []types.Dependency, err error) {
 	scanner := bufio.NewScanner(r)
 	unique := map[string]struct{}{}
-	var lib types.Library
+	dependsOn := map[string][]Dependency{}
+	yarnLibs := map[string]Library{}
+	var lib Library
 	var skipPackage bool
+	var isInPackage bool
+	var inDependenciesBlock bool
 	var lineNumber int // It is used to save dependency location
 	for scanner.Scan() {
 		lineNumber++
 		line := scanner.Text()
 		if len(line) < 1 {
+			// save previous package
+			if isInPackage {
+				isInPackage = false
+				inDependenciesBlock = false
+				// fetch between version prefix and last double-quote
+				symbol := utils.PackageID(lib.Name, lib.Version)
+				if _, ok := unique[symbol]; ok {
+					lib = Library{}
+					continue
+				}
+
+				lib.Location.EndLine = lineNumber - 1
+
+				for _, loc := range lib.Locators {
+					yarnLibs[loc] = lib
+				}
+				unique[symbol] = struct{}{}
+
+				lib = Library{}
+			}
 			continue
 		}
 
+		// parse dependency
+		if inDependenciesBlock {
+			if name, version, err := getDependency(line); err == nil {
+				dep := Dependency{
+					Locator: utils.PackageID(name, version),
+					Name:    name,
+				}
+				lo.ForEach(lib.Locators, func(loc string, _ int) {
+					if _, ok := dependsOn[loc]; !ok {
+						dependsOn[loc] = []Dependency{}
+					}
+					dependsOn[loc] = append(dependsOn[loc], dep)
+				})
+				continue
+			} else {
+				inDependenciesBlock = false
+			}
+		}
+
 		// parse version
-		var version string
-		if version, err = getVersion(line); err == nil {
+		if version, err := getVersion(line); err == nil {
 			if skipPackage {
 				continue
 			}
+
 			if lib.Name == "" {
 				return nil, nil, xerrors.New("Invalid yarn.lock format")
 			}
-			// fetch between version prefix and last double-quote
-			symbol := fmt.Sprintf("%s@%s", lib.Name, version)
-			if _, ok := unique[symbol]; ok {
-				lib = types.Library{}
-				continue
-			}
 
 			lib.Version = version
-			libs = append(libs, lib)
-			lib = types.Library{}
-			unique[symbol] = struct{}{}
 			continue
 		}
+
 		// skip __metadata block
 		if skipPackage = strings.HasPrefix(line, "__metadata"); skipPackage {
 			continue
 		}
+
 		// packagename line start 1 char
 		if line[:1] != " " && line[:1] != "#" {
```

Comment:
There are now many states like `isInPackage`, `inDependenciesBlock` and `skipPackage` in a big scope, `Parse`. What if parsing blocks first and then parsing each line instead of having the states?

```
// scanBlocks splits "yarn.lock" by blocks.
func scanBlocks(data []byte, atEOF bool) (advance int, token []byte, err error) {
	if atEOF && len(data) == 0 {
		return 0, nil, nil
	}
	if i := bytes.Index(data, []byte("\n\n")); i >= 0 {
		// We have a full newline-terminated line.
		return i + 2, data[0:i], nil
	} else if i := bytes.Index(data, []byte("\r\n\r\n")); i >= 0 {
		return i + 4, data[0:i], nil
	}

	// If we're at EOF, we have a final, non-terminated line. Return it.
	if atEOF {
		return len(data), data, nil
	}
	// Request more data.
	return 0, nil, nil
}

func (p *Parser) Parse(r dio.ReadSeekerAt) (libs []types.Library, deps []types.Dependency, err error) {
	scanner := bufio.NewScanner(r)
	scanner.Split(scanBlocks) // Split by package blcoks
	for scanner.Scan() {
		block := scanner.Bytes()
		pkg, err := parseBlock(block)
		...
	}
	...
}

func parseBlock(block []byte) {
	scanner := bufio.NewScanner(bytes.NewReader(block))
	var pkg types.Package
	for scanner.Scan() {
		line := scanner.Text()

		// Skip this block
		if strings.HasPrefix(line, "__metadata") {
			return
		}

		// The first line in this package block
		if if line[:1] != " " && line[:1] != "#" {
			parsePackageLocators(line)
			continue
		}

		line = strings.TrimSpace(line)
		switch {
		case strings.HasPrefix(line, "version"):
			pkg.Version, err = parseVersion(line)
		case line == "dependencies:" {
			// start dependencies block
			deps, err := parseDependencies(scanner)
		}
	}
	return pkg, ..., err
}

func parseDependencies(scanner *bufio.Scanner) []types.Dependency {
	var deps []types.Dependency
	for scanner.Scan() {
		line := scanner.Text()
		parseDependency(line)
		...
	}
	return deps
}
```

What do you think of the idea? We call `bufio.NewScanner` twice for splitting blocks and lines, respectively. Then, we don't need to reset those states like `lib = Library{}`, `isInPackage = false` and `inDependenciesBlock = false`. I don't think the above code works, but hopefully, you understand my idea.

We may want to define `ScannerWithLine` to keep the current line number, but it is no big deal. We can manage it somehow.